File: 백준/Silver/14716. 현수막/현수막.java
import java.io.*;
import java.util.*;

public class Main {
	static int[] dirX = {-1,-1,0,1,1,1,0,-1};	//상 상우 우 우하 하 좌하 좌 좌상
	static int[] dirY = {0,1,1,1,0,-1,-1,-1};
	
	static int N, M;
	static int[][] map;
	
	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		map = new int[N][M];
		
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < M; j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		int cnt = 0;

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if(map[i][j] == 1) {
					bfs(i, j);
					cnt++;
				}
			}
		}
		
		System.out.println(cnt);
		br.close();
	}

	private static void bfs(int x, int y) {
		ArrayDeque<int[]> q = new ArrayDeque<>();
		
		q.offer(new int[] {x, y});
		
		while (!q.isEmpty()) {
			int[] poll = q.poll();
			
			for (int d = 0; d < 8; d++) {
				int dx = poll[0] + dirX[d];
				int dy = poll[1] + dirY[d];
				
				if(0<=dx && dx<N && 0<=dy && dy<M && map[dx][dy]==1) {
					map[dx][dy] = 0;
					q.offer(new int[] {dx, dy});
				}
			}
		}
	}
}

File: 백준/Silver/14716. 현수막/현수막.py
dirX = [-1,-1,0,1,1,1,0,-1]   # 상 상우 우 우하 하 하좌 좌 좌상
dirY = [0,1,1,1,0,-1,-1,-1]

import sys
input = sys.stdin.readline

N, M = map(int, input().split())

banner = [[] for _ in range(N)]

for i in range(N) :
    banner[i] = list(map(int,input().split()))

def bfs(i, j) :
    q = []
    q.append([i, j])

    while len(q) != 0 :
        pop = q.pop()

        for d in range(8) :
            dx = pop[0] + dirX[d]
            dy = pop[1] + dirY[d]

            if 0<=dx and dx<N and 0<=dy and dy<M :
                if banner[dx][dy] == 0 :
                    continue
                banner[dx][dy] = 0
                q.append([dx,dy])

cnt = 0

for i in range(N) :
    for j in range(M) :
        if banner[i][j] == 1 :
            bfs(i,j)
            cnt += 1

print(cnt)